src/renderer.c

```diff
@@ -137,97 +175,195 @@ static int font_set_style(FT_Outline* outline, int x_translation, unsigned char
-      if (FT_Render_Glyph(slot, render_option))
-        continue;
-      for (unsigned int line = 0; line < slot->bitmap.rows; ++line) {
-        int target_offset = set->surface->pitch * line + set->metrics[i].x0 * byte_width;
-        int source_offset = line * slot->bitmap.pitch;
-        if (font->antialiasing == FONT_ANTIALIASING_NONE) {
-          for (unsigned int column = 0; column < slot->bitmap.width; ++column) {
-            int current_source_offset = source_offset + (column / 8);
-            int source_pixel = slot->bitmap.buffer[current_source_offset];
-            pixels[++target_offset] = ((source_pixel >> (7 - (column % 8))) & 0x1) * 0xFF;
-          }
-        } else
-          memcpy(&pixels[target_offset], &slot->bitmap.buffer[source_offset], slot->bitmap.width);
+  }
+  if (atlas_idx < 0) {
+    // create a new atlas with the correct width, for each subpixel bitmap  
+    for (int i = 0; i < FONT_BITMAP_COUNT(font); i++) {
+      font->glyphs.atlas[i] = check_alloc(realloc(font->glyphs.atlas[i], sizeof(GlyphAtlas) * (font->glyphs.natlas + 1)));
+      font->glyphs.atlas[i][font->glyphs.natlas] = (GlyphAtlas) {
+        .width = metric->x1 + FONT_WIDTH_OVERFLOW_PX, .nsurface = 0,
+        .surfaces = NULL,
+      };
+    }
+    font->glyphs.bytesize += sizeof(GlyphAtlas);
+    atlas_idx = font->glyphs.natlas++;
+  }
+  metric->atlas_idx = atlas_idx;
+  GlyphAtlas *atlas = &font->glyphs.atlas[bitmap_idx][atlas_idx];
+
+  // find the surface with the minimum width that can fit the glyph (limited to last 100 surfaces)
+  int surface_idx = -1, max_surface_idx = (int) atlas->nsurface - 100, min_waste = INT_MAX;
+  for (int i = atlas->nsurface - 1; i >= 0 && i > max_surface_idx; i--) {
+    int new_min_waste = (int) atlas->surfaces[i]->h - (intptr_t) atlas->surfaces[i]->userdata;
+    if (new_min_waste >= 0 && new_min_waste < min_waste) {
+      surface_idx = i;
+      min_waste = new_min_waste;
+    }
+  }
+  if (surface_idx < 0) {
+    // allocate a new surface array, and a surface
+    int h = FONT_HEIGHT_OVERFLOW_PX + (double) font->face->size->metrics.height / 64.0f;
+    if (h <= FONT_HEIGHT_OVERFLOW_PX) h += slot->bitmap.rows;
+    if (h <= FONT_HEIGHT_OVERFLOW_PX) h += font->size;
+    atlas->surfaces = check_alloc(realloc(atlas->surfaces, sizeof(SDL_Surface *) * (atlas->nsurface + 1)));
+    atlas->surfaces[atlas->nsurface] = check_alloc(SDL_CreateRGBSurface(
+      0, atlas->width, GLYPH_PER_ATLAS * h, FONT_BITMAP_COUNT(font) * 8,
+      0, 0, 0, 0
+    ));
+    atlas->surfaces[atlas->nsurface]->userdata = 0;
+    surface_idx = atlas->nsurface++;
+    font->glyphs.bytesize += (sizeof(SDL_Surface *) + sizeof(SDL_Surface) + atlas->width * GLYPH_PER_ATLAS * h * FONT_BITMAP_COUNT(font));
+  }
+  metric->surface_idx = surface_idx;
+  metric->y0 = (uintptr_t) atlas->surfaces[surface_idx]->userdata;
+  atlas->surfaces[surface_idx]->userdata = (void *) ((uintptr_t) atlas->surfaces[surface_idx]->userdata + metric->y1);
```

Comment:
Would it make sense to point `userdata` to `metric`?
So that the `userdata` contains the last glyph added to a surface, and you don't have to check that the value fits inside a pointer.

Reply:
Good point. Didn't think of that.